src/modules/control_allocator/ControlAllocation/ControlAllocation.cpp

```diff
@@ -0,0 +1,150 @@
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+/**
+ * @file ControlAllocation.cpp
+ *
+ * Interface for Control Allocation Algorithms
+ *
+ * @author Julien Lecoeur <julien.lecoeur@gmail.com>
+ */
+
+#include "ControlAllocation.hpp"
+
+const matrix::Vector<float, ControlAllocation::NUM_ACTUATORS> &
+ControlAllocation::getAllocatedActuator() const
+{
+	return _actuator_sp;
+}
+
+void
+ControlAllocation::setControlSetpoint(const matrix::Vector<float, ControlAllocation::NUM_AXES> &control)
+{
+	_control_sp = control;
+}
+
+const matrix::Vector<float, ControlAllocation::NUM_AXES> &
+ControlAllocation::getControlSetpoint() const
+{
+	return _control_sp;
+}
+
+const matrix::Vector<float, ControlAllocation::NUM_AXES> &
+ControlAllocation::getAllocatedControl() const
+{
+	return _control_allocated;
+}
+
+void
+ControlAllocation::setEffectivenessMatrix(const
+		matrix::Matrix<float, ControlAllocation::NUM_AXES, ControlAllocation::NUM_ACTUATORS> &B)
+{
+	_B = B;
+}
+
+const matrix::Matrix<float, ControlAllocation::NUM_AXES, ControlAllocation::NUM_ACTUATORS> &
+ControlAllocation::getEffectivenessMatrix() const
+{
+	return _B;
+}
+
+void
+ControlAllocation::setActuatorMin(const matrix::Vector<float, ControlAllocation::NUM_ACTUATORS>
+				  &actuator_min)
+{
+	_actuator_min = actuator_min;
+}
+
+const matrix::Vector<float, ControlAllocation::NUM_ACTUATORS> &
+ControlAllocation::getActuatorMin() const
+{
+	return _actuator_min;
+}
+
+void
+ControlAllocation::setActuatorMax(const matrix::Vector<float, ControlAllocation::NUM_ACTUATORS>
+				  &actuator_max)
+{
+	_actuator_max = actuator_max;
+}
+
+const matrix::Vector<float, ControlAllocation::NUM_ACTUATORS> &
+ControlAllocation::getActuatorMax() const
+{
+	return _actuator_max;
+}
+
+void
+ControlAllocation::setCurrentActuatorSetpoint(const matrix::Vector<float, ControlAllocation::NUM_ACTUATORS>
+		&actuator_sp)
+{
+	_actuator_sp = actuator_sp;
+}
+
+matrix::Vector<float, ControlAllocation::NUM_ACTUATORS>
+ControlAllocation::clipActuatorSetpoint() const
+{
+	matrix::Vector<float, ControlAllocation::NUM_ACTUATORS> actuator_clipped = _actuator_sp;
+
+	for (size_t i = 0; i < ControlAllocation::NUM_ACTUATORS; i++) {
+		if (_actuator_max(i) < _actuator_min(i)) {
+			// this should not happen
+			actuator_clipped(i) = _actuator_min(i);
```

Comment:
@jlecoeur Is this how you were intending to deliberately enforce a specific value for an actuator?

Reply:
Yes that is one way. It disables completely the actuator (fixed value + excluded from the allocation).

Another way I was considering is to add a "trim" variable, so that the actuator works around a specific value, but is not excluded from the allocation. Something like:
$$
u = u_0 + B^{+} \cdot (c - c_0)
$$
with u the actuator values, u_0 the actuator trims, c the control setpoint, and c_0 = B u_0, and B the effectiveness around u_0.

For example with a tiltrotor, the trim associated with the tilting actuators is modified during the transition, so the rotors are "forced" to tilt but can still be used for control. Meanwhile, the effectiveness matrix updated to match the current trim.

I was intending to use the same implementation for effectiveness as for allocation: i.e. a pointer to an "EffectivenessProvider" class that provides effectiveness, trims, etc